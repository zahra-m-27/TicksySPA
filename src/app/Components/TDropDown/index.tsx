import {useEffect, useRef, useState} from 'react';
import Assets from '../../Assets';
import styles from './styles.module.scss';
import ClassNames from '../../Utilities/ClassNames';

interface Item<T> {
  value: T;
  label: string;
}

interface Props<T> {
  label?: string;
  items: Item<T>[];
  className?: string;
  selectedItem?: T[];
  multiSelect?: boolean;
  labelClassName?: string;
  onSelect?: (values: T[]) => void;
}

export default function TDropDown<T>({
  label,
  items,
  onSelect,
  className,
  multiSelect,
  selectedItem,
  labelClassName,
}: Props<T>) {
  const [Open, setOpen] = useState<boolean>(false);
  const container = useRef<HTMLDivElement>(null);
  const input = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const onClickContainer = () => setOpen(false);
    const onClickInput = (e: any) => {
      e.stopPropagation();
      setOpen(!Open);
    };
    container.current?.parentElement?.addEventListener(
      'click',
      onClickContainer,
      false
    );
    input.current?.addEventListener('click', onClickInput, false);
    return () => {
      container.current?.parentElement?.removeEventListener(
        'click',
        onClickContainer
      );
      input.current?.removeEventListener('click', onClickInput);
    };
  }, [Open]);

  const label_style = ClassNames(styles.label, labelClassName);
  const container_style = ClassNames(styles.container, className);

  const selectedLabel = items
    ?.filter((i) => selectedItem?.includes(i.value))
    .map((i) => i.label)
    ?.join(', ');

  return (
    <div ref={container} data-open={Open} className={container_style}>
      <p className={label_style}>{label}</p>
      <div
        ref={input}
        className={styles.content}
        data-testid="dropdown-container">
        <img
          alt="arrow down"
          src={Assets.SVGs.ArrowDown}
          className={styles.arrow_down}
        />
        <p className={styles.selected_item_label}>
          {selectedLabel.length ? selectedLabel : 'انتخاب كنيد'}
        </p>
      </div>

      <div className={styles.items}>
        {items.map((item) => (
          <div
            key={`${item.value}`}
            className={styles.item}
            onClick={() => {
              setOpen(!!multiSelect);
              if (selectedItem) {
                const index = selectedItem.indexOf(item.value);
                if (index >= 0) {
                  onSelect &&
                    onSelect(selectedItem.filter((v, i) => i != index));
                } else {
                  onSelect && onSelect([item.value, ...selectedItem]);
                }
              } else {
                onSelect && onSelect([item.value]);
              }
            }}>
            {multiSelect &&
              selectedItem &&
              selectedItem.indexOf(item.value) >= 0 && (
                <Assets.SVGs.CheckmarkSVG className={styles.checkmark} />
              )}
            <p>{item.label}</p>
          </div>
        ))}
      </div>
    </div>
  );
}
