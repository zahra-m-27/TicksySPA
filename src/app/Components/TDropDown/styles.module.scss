.container {
  width: 100%;
  cursor: pointer;
  text-align: end;
  margin-top: 10px;
  position: relative;

  .label {
    color: #333333;
    margin-bottom: 10px;
  }

  .content {
    display: flex;
    padding: 10px 15px;
    border-radius: 5px;
    background-color: #dadde0;
    border: 1px solid #e3e3e3;

    p {
      flex: 1;
      margin: 0;
      text-align: end;
    }
  }

  .arrow_down {
  }

  &[data-open='false'] {
    .items {
      display: none;
    }
  }

  .selected_item_label {
    color: #606f7b;
    cursor: pointer;
  }

  .items {
    width: 100%;
    padding: 10px;
    cursor: pointer;
    transition: 0.6s;
    position: absolute;
    border-radius: 8px;
    background-color: #eeeff1;
    box-shadow: 0 3px 6px 0 rgb(0 0 0 / 16%);

    .item {
      color: #000;
      margin: 0.5px;
      display: block;
      text-align: end;
      padding: 8px 20px;

      &:hover {
        overflow: hidden;
        position: relative;
        border-radius: 5px;
        background-color: #dadde0;
      }
    }
  }
}
