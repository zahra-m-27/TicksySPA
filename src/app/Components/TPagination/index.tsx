import React from 'react';
import styles from './styles.module.scss';
import Assets from '../../Assets';
import ClassNames from '../../Utilities/ClassNames';

interface Props {
  total: number;
  pageSize: number;
  className?: string;
  pageNumber: number;
  onChange: (pageNumber: number) => void;
}

export default function TPagination({
  pageNumber,
  className,
  pageSize,
  onChange,
  total,
}: Props) {
  const lastPage = Math.ceil(total / pageSize);

  return (
    <div className={ClassNames(styles.pagination, className)}>
      <Assets.SVGs.MoreThan
        className={styles.move_button}
        onClick={() => pageNumber - 1 > 0 && onChange(pageNumber - 1)}
      />
      <div className={styles.divider} />
      {pageNumber - 1 > 0 && (
        <div
          className={styles.pagination_page}
          onClick={() => onChange(pageNumber - 1)}>
          {pageNumber - 1}
        </div>
      )}
      <div className={ClassNames(styles.pagination_page, styles.active)}>
        {pageNumber}
      </div>
      {pageNumber + 1 <= lastPage && (
        <div
          className={styles.pagination_page}
          onClick={() => onChange(pageNumber + 1)}>
          {pageNumber + 1}
        </div>
      )}
      <div className={styles.divider} />
      <Assets.SVGs.LessThan
        className={styles.move_button}
        onClick={() => pageNumber + 1 <= lastPage && onChange(pageNumber + 1)}
      />
    </div>
  );
}
