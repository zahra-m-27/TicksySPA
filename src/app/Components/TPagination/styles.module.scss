.pagination {
  display: flex;
  direction: rtl;
  align-items: center;
  justify-content: center;

  .divider {
    width: 3px;
    height: 43px;
    margin: 0 15px;
    border-radius: 1px;
    background-color: #e8e8e8;
  }

  .pagination_page {
    width: 40px;
    height: 40px;
    display: flex;
    margin: 0 5px;
    font-size: 17px;
    cursor: pointer;
    user-select: none;
    border-radius: 5px;
    align-items: center;
    justify-content: center;

    &.active {
      color: white;
      background-color: #44adf2;
    }
  }

  .move_button {
    cursor: pointer;
  }
}
