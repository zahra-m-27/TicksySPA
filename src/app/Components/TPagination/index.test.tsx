import TDropDown from './index';
import React from 'react';
import {render} from '@testing-library/react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import user from '@testing-library/user-event';
import TPagination from './index';

const onChange = jest.fn();

const sample = (
  <TPagination total={100} pageSize={20} pageNumber={2} onChange={onChange} />
);

test('render in ReactDOM', () => {
  const div = document.createElement('div');
  ReactDOM.render(sample, div);
});

test('render items', () => {
  const {getByText} = render(sample);
  getByText(/1/i);
  getByText(/2/i);
  getByText(/3/i);
});

test('on page click', () => {
  const {getByText} = render(sample);
  const item1 = getByText(/1/i);
  user.click(item1);
  expect(onChange).toBeCalledTimes(1);
  expect(onChange).toBeCalledWith(1);

  const item3 = getByText(/3/i);
  user.click(item3);
  expect(onChange).toBeCalledTimes(2);
  expect(onChange).toBeCalledWith(3);
});

test('on arrow click', () => {
  const {getByTestId} = render(sample);
  const previous = getByTestId('previous_page_arrow');
  user.click(previous);
  expect(onChange).toBeCalledTimes(1);
  expect(onChange).toBeCalledWith(1);

  const next = getByTestId('next_page_arrow');
  user.click(next);
  expect(onChange).toBeCalledTimes(2);
  expect(onChange).toBeCalledWith(3);
});

it('snapshot', () => {
  const snapshot = renderer.create(sample).toJSON();
  expect(snapshot).toMatchSnapshot();
});
