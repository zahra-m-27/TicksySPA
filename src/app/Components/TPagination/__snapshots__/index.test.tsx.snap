// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  className="pagination"
>
  <svg
    className="move_button"
    data-testid="previous_page_arrow"
    height="23.889"
    onClick={[Function]}
    viewBox="0 0 13.98 23.889"
    width="13.98"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M19.067 8.313l-2.035 2.036 9.931 9.909-9.931 9.931 2.012 2.011 10.938-10.936 1.029-1.007-1.029-1.007z"
      fill="#76787d"
      transform="translate(-17.031 -8.313)"
    />
  </svg>
  <div
    className="divider"
  />
  <div
    className="pagination_page"
    onClick={[Function]}
  >
    1
  </div>
  <div
    className="pagination_page active"
  >
    2
  </div>
  <div
    className="pagination_page"
    onClick={[Function]}
  >
    3
  </div>
  <div
    className="divider"
  />
  <svg
    className="move_button"
    data-testid="next_page_arrow"
    height="23.889"
    onClick={[Function]}
    viewBox="0 0 13.98 23.889"
    width="13.98"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M28.976 8.313l2.035 2.036-9.931 9.908 9.931 9.931L29 32.2 18.06 21.264l-1.029-1.007 1.029-1.007z"
      fill="#76787d"
      transform="translate(-17.031 -8.313)"
    />
  </svg>
</div>
`;
